control-plane/stor-port/src/types/v0/transport/nexus.rs

```diff
@@ -79,6 +79,101 @@ impl From<Nexus> for models::Nexus {
 
 rpc_impl_string_uuid!(NexusId, "UUID of a nexus");
 
+/// A gRPC request type for creating snapshot of a nexus, which essentially
+/// means a snapshot of all(or selected) replicas associated with that nexus.
+pub struct CreateNexusSnapshot {
+    /// uuid of the nexus.
+    pub nexus: NexusId,
+    /// uuid of the entity, nexus in this case.
+    pub entity: uuid::Uuid,
+    /// A transaction id for this request.
+    pub txn_id: String,
+    /// name of the snapshot to be created.
+    pub name: String,
+    /// replica list descriptor.
+    pub replica_desc: Vec<CreateNexusSnapReplDescr>,
+}
+
+/// A descriptor that specifies the replicas which need to taken snapshot
+/// of specifically.
+#[derive(Clone)]
+#[allow(unused)]
+pub struct CreateNexusSnapReplDescr {
+    pub replica: ReplicaId,
+    pub snap_uuid: Option<SnapshotId>,
+    pub skip: bool,
+}
+
+impl CreateNexusSnapshot {
+    /// Create a new request.
+    pub fn new(
+        nexus: &NexusId,
+        entity: uuid::Uuid,
+        txn_id: String,
+        name: String,
+        replica_desc: Vec<CreateNexusSnapReplDescr>,
+    ) -> Self {
+        Self {
+            nexus: nexus.clone(),
+            entity,
+            txn_id,
+            name,
+            replica_desc,
+        }
+    }
+    /// Get a reference to the nexus uuid.
+    pub fn nexus(&self) -> &NexusId {
+        &self.nexus
+    }
+    /// Get a reference to the entity id.
+    pub fn entity(&self) -> &uuid::Uuid {
+        &self.entity
+    }
+    /// Get a reference to the transaction id.
+    pub fn txn_id(&self) -> &str {
+        &self.txn_id
+    }
+    /// Get a reference to the snapshot name.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+    /// Get a reference to the replica descriptor.
+    pub fn replica_desc(&self) -> &Vec<CreateNexusSnapReplDescr> {
+        &self.replica_desc
+    }
+}
+
+pub struct CreateNexusSnapshotResp {
+    /// uuid of the nexus.
+    pub nexus: NexusId,
```

Comment:
should these be private?

Reply:
for the simpler structs I haven't implemented impl , and since they'll be required in translation kept them public.